typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
-        name,
-        mediaType,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
-    case Opcode.STATISTICS: {
-      const messageCount = reader.uint64();
-      const schemaCount = reader.uint16();
-      const channelCount = reader.uint32();
-      const attachmentCount = reader.uint32();
-      const metadataCount = reader.uint32();
-      const chunkCount = reader.uint32();
-      const messageStartTime = reader.uint64();
-      const messageEndTime = reader.uint64();
-      const channelMessageCounts = reader.map(
-        (r) => r.uint16(),
-        (r) => r.uint64(),
-      );
-
-      const record: TypedMcapRecord = {
-        type: "Statistics",
-        messageCount,
-        schemaCount,
-        channelCount,
-        attachmentCount,
-        metadataCount,
-        chunkCount,
-        messageStartTime,
-        messageEndTime,
-        channelMessageCounts,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
-    case Opcode.METADATA: {
-      const name = reader.string();
-      const metadata = reader.map(
-        (r) => r.string(),
-        (r) => r.string(),
-      );
-      const record: TypedMcapRecord = { type: "Metadata", metadata, name };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
-    case Opcode.METADATA_INDEX: {
-      const offset = reader.uint64();
-      const length = reader.uint64();
-      const name = reader.string();
-
-      const record: TypedMcapRecord = {
-        type: "MetadataIndex",
-        offset,
-        length,
-        name,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
-    case Opcode.SUMMARY_OFFSET: {
-      const groupOpcode = reader.uint8();
-      const groupStart = reader.uint64();
-      const groupLength = reader.uint64();
-
-      const record: TypedMcapRecord = {
-        type: "SummaryOffset",
-        groupOpcode,
-        groupStart,
-        groupLength,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
-    case Opcode.DATA_END: {
-      const dataSectionCrc = reader.uint32();
-      const record: TypedMcapRecord = {
-        type: "DataEnd",
-        dataSectionCrc,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
+function parseAttachment(
+  reader: Reader,
+  recordLength: number,
+  // eslint-disable-next-line @foxglove/no-boolean-parameters
+  validateCrcs: boolean,
+): TypedMcapRecord {
+  const startOffset = reader.offset;
+  const logTime = reader.uint64();
+  const createTime = reader.uint64();
+  const name = reader.string();
+  const mediaType = reader.string();
+  const dataLen = reader.uint64();
+  // NOTE: probably not necessary, but just in case
+  if (BigInt(reader.offset) + dataLen > Number.MAX_SAFE_INTEGER) {
+    throw new Error(`Attachment too large: ${dataLen}`);
+  }
+  if (reader.offset + Number(dataLen) + 4 /*crc*/ > startOffset + recordLength) {
```

Comment:
similar to above comment, it feels like we could now rely on u8ArrayCopy() for the bounds check

Reply:
As explained above, best to move error throwing up not down the tree